File: BSIGeneralAffairDAL/DALOffice.vb
Imports System.Data
Imports System.Data.SqlClient
Imports BSIGeneralAffairBO
Imports GeneralAffairInterface

Public Class DALOffice
    Implements IOffice

    Public database As DatabaseConfiguration = New DatabaseConfiguration()
    Private dataRead As SqlDataReader
    Public cmd As SqlCommand

    Public Function create(obj As Office) As Integer Implements IBaseCRUD(Of Office).create
        Try
            Dim strSP = "[Office].[USP_StoreOffice]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure


            cmd.Parameters.AddWithValue("@OfficeName", obj.OfficeName)
            cmd.Parameters.AddWithValue("@OfficeAddress", obj.OfficeAddress)
            cmd.Parameters.AddWithValue("@OfficeFlagActive", obj.OfficeFlagActive)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            If result = 1 Then
                Throw New ArgumentException("Office not created")
            End If
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function getAll() As List(Of Office) Implements IBaseCRUD(Of Office).getAll
        Dim Offices As New List(Of Office)
        Try
            Dim strSP = "[Office].[USP_SelectAllOffice]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                While dataRead.Read
                    Dim office As New Office
                    office.OfficeID = CInt(dataRead("OfficeID"))
                    office.OfficeName = dataRead("OfficeName").ToString()
                    office.OfficeAddress = dataRead("OfficeAddress").ToString()
                    office.OfficeFlagActive = CInt(dataRead("OfficeFlagActive"))
                    Offices.Add(office)
                End While
            End If
            dataRead.Close()

            Return Offices
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function update(obj As Office) As Integer Implements IBaseCRUD(Of Office).update
        Try
            Dim updateOffice = GetById(obj.OfficeID)
            If updateOffice Is Nothing Then
                Throw New ArgumentException("Office not found")
            End If

            Dim strSP = "[Office].[USP_UpdateOffice]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            cmd.Parameters.AddWithValue("@OfficeID", obj.OfficeID)
            cmd.Parameters.AddWithValue("@OfficeName", obj.OfficeName)
            cmd.Parameters.AddWithValue("@OfficeAddress", obj.OfficeAddress)
            cmd.Parameters.AddWithValue("@OfficeFlagActive", obj.OfficeFlagActive)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
            Return result
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Private Function GetById(officeID As Integer) As Object
        Try
            cmd = New SqlCommand("SELECT * FROM [Office].[OfficeLocation] WHERE OfficeName = @OfficeName", database.conn)
            cmd.Parameters.AddWithValue("@OfficeName", officeID)
            database.conn.Open()
            dataRead = cmd.ExecuteReader()
            If dataRead.HasRows Then
                dataRead.Read()
                Dim office As New Office
                office.OfficeID = CInt(dataRead("OfficeID"))
                office.OfficeName = dataRead("OfficeName ").ToString()
                office.OfficeName = dataRead("OfficeName ").ToString()
                office.UpdateAt = dataRead("UpdatedAt").ToString()
                Return office
            Else
                Return Nothing
            End If
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function

    Public Function delete(id As Integer) As Integer Implements IBaseCRUD(Of Office).delete
        Try
            Dim office = GetById(id)
            If office Is Nothing Then
                Throw New ArgumentException("Office not found")
            End If

            Dim strSP = "[Office].[USP_DeleteOffice]"
            cmd = New SqlCommand(strSP, database.conn)
            cmd.CommandType = CommandType.StoredProcedure
            cmd.Parameters.AddWithValue("@OfficeID", id)

            database.conn.Open()
            Dim result = cmd.ExecuteNonQuery()
        Catch sqlex As SqlException
            Throw New ArgumentException(sqlex.Message & " " & sqlex.Number)
        Catch ex As Exception
            Throw ex
        Finally
            cmd.Dispose()
            database.conn.Close()
        End Try
    End Function
End Class

File: BSIGeneralAffairDAL-C/DALOffice.cs
using BSIGeneralAffairBO_C;
using BSIGeneralAffairDAL_C.Interfaces;
using Dapper;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SqlClient;
using System.Text;

namespace BSIGeneralAffairDAL_C
{
    public class DALOffice : IOfficeDAL
    {
        public void Delete(int id)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"[Office].[USP_DeleteOffice]";
                var param = new { OfficeID = id };
                try
                {
                    int result = conn.Execute(strSql, param);
                    if (result == 1)
                    {
                        throw new ArgumentException("Delete data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }
        }

        private string GetConnectionString()
        {
            return ConfigurationManager.ConnectionStrings["MyDbConnectionString"].ConnectionString;
        }

        public IEnumerable<Office> GetAll()
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                //var strSql = @"select * from Categories order by CategoryName";
                var strSql = @"[Office].[USP_SelectAllOffice]";

                var results = conn.Query<Office>(strSql, commandType: System.Data.CommandType.StoredProcedure);
                return results;
            }
        }

        public Office GetById(int id)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [Office].[OfficeLocation] Where [OfficeID] = @OfficeID ";
                var param = new { OfficeID = id };
                var result = conn.QuerySingleOrDefault<Office>(strSql, param);
                return result;
            }
        }

        public IEnumerable<Office> GetByName(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [Office].[OfficeLocation] Where [OfficeName] = @OfficeName ";
                var param = new { OfficeName = name };
                var result = conn.Query<Office>(strSql, param);
                return result;
            }
        }

        public int GetCountOffices(string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT COUNT(*) FROM [Office].[OfficeLocation] Where [OfficeName] like @OfficeName ";
                var param = new { OfficeName = $"%{name}%" };
                var result = Convert.ToInt32(conn.ExecuteScalar(strSql, param));
                return result;
            }
        }

        public IEnumerable<Office> GetWithPaging(int pageNumber, int pageSize, string name)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = @"SELECT * FROM [Office].[OfficeLocation] 
                              Where [OfficeName] like @OfficeName 
                              order by OfficeName OFFSET @Offset ROWS FETCH NEXT @PageSize ROWS ONLY";
                var param = new { OfficeName = $"%{name}%", Offset = (pageNumber - 1) * pageSize, PageSize = pageSize };
                var results = conn.Query<Office>(strSql, param);
                return results;
            }
        }

        public void Insert(Office entity)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                var strSql = "[Office].[USP_StoreOffice]";
                var param = new { 
                    OfficeName = entity.OfficeName,
                    OfficeAddress = entity.OfficeAddress,
                    OfficeFlagActive = entity.OfficeFlagActive
                };
                try
                {
                    int result = conn.Execute(strSql, param, commandType: System.Data.CommandType.StoredProcedure);
                    if (result == 1)
                    {
                        throw new ArgumentException("Insert data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new ArgumentException($"{sqlEx.InnerException.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new ArgumentException(ex.Message);
                }
            }
        }

        public void Update(Office entity)
        {
            using (SqlConnection conn = new SqlConnection(GetConnectionString()))
            {
                try
                {
                    var strSql = @"[Office].[USP_UpdateOffice]";
                    var param = new
                    {
                        OfficeID = entity.OfficeID,
                        OfficeName = entity.OfficeName,
                        OfficeAddress = entity.OfficeAddress,
                        OfficeFlagActive = entity.OfficeFlagActive
                    };
                    int result = conn.Execute(strSql, param);

                    //jika result = -1, berarti update data gagal
                    if (result == 1)
                    {
                        throw new Exception("Update data failed..");
                    }
                }
                catch (SqlException sqlEx)
                {
                    throw new Exception($"{sqlEx.Message} - {sqlEx.Number}");
                }
                catch (Exception ex)
                {
                    throw new Exception(ex.Message);
                }
            }
        }
    }
}
